client/wsclient.go

```diff
@@ -245,19 +250,33 @@ func (c *wsClient) runOneCycle(ctx context.Context) {
 		c.common.PackagesStateProvider,
 		c.common.Capabilities,
 	)
-	r.ReceiverLoop(ctx)
-
-	// Stop the background processors.
-	procCancel()
-
-	// If we exited receiverLoop it means there is a connection error, we cannot
-	// read messages anymore. We need to start over.
+	r.Start(ctx)
+
+	select {
+	case <-c.sender.IsStopped():
+		// sender will send close message to initiate the close handshake
+		if err := c.sender.Err(); err != nil && err != websocket.ErrCloseSent {
+			c.common.Logger.Debugf("failed to send close message, close without the handshake.")
+			break
+		}
 
-	// Close the connection to unblock the WSSender as well.
-	_ = c.conn.Close()
+		c.common.Logger.Debugf("waiting for close message from server.")
+		select {
+		case <-r.IsStopped():
+			c.common.Logger.Debugf("shutdown handshake complete.")
+		case <-time.After(c.connShutdownTimeout):
+			c.common.Logger.Debugf("timeout waiting for close message.")
+			// not receive close message from the server, close the connection to force the receive loop to stop
+			_ = c.conn.Close()
+			<-r.IsStopped()
```

Comment:
Can this potentially block forever? To be defensive we probably need to have a timeout here.

Reply:
Receiver is waiting to read data from the connection (via `ReadMessage`), I think closing the connection is enough to unblock the `ReadMessage`. 🤔